src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/InterpolatedStringBuilder.cs

```diff
@@ -0,0 +1,596 @@
+        //   for alignment and format, etc.
+        //
+        // To address that, we could just have overloads for T and ReadOnlySpan<char>:
+        //     (T)
+        //     (T, int alignment)
+        //     (T, string? format)
+        //     (T, int alignment, string? format)
+        //     (ReadOnlySpan<char>)
+        //     (ReadOnlySpan<char>, int alignment)
+        //     (ReadOnlySpan<char>, string? format)
+        //     (ReadOnlySpan<char>, int alignment, string? format)
+        // but this also has shortcomings:
+        // - Some expressions that would have worked with an object overload will now force a fallback to string.Format
+        //   (or fail to compile if the builder is used in places where the fallback isn't provided), because the compiler
+        //   can't always target type to T, e.g. `b switch { true => 1, false => null }` where `b` is a bool can successfully
+        //   be passed as an argument of type `object` but not of type `T`.
+        // - Reference types get no benefit from going through the generic code paths, and actually incur some overheads
+        //   from doing so.
+        // - Nullable value types also pay a heavy price, in particular around interface checks that would generally evaporate
+        //   at compile time for value types but don't (currently) if the Nullable<T> goes through the same code paths
+        //   (see https://github.com/dotnet/runtime/issues/50915).
+        //
+        // We could try to take a more elaborate approach for InterpolatedStringBuilder, since it is the most common builder
+        // and we want to minimize overheads both at runtime and in IL size, e.g. have a complete set of overloads for each of:
+        //     (T, ...) where T : struct
+        //     (T?, ...) where T : struct
+        //     (object, ...)
+        //     (ReadOnlySpan<char>, ...)
+        //     (string, ...)
+        // but this also has shortcomings, most importantly:
+        // - If you have an unconstrained T that happens to be a value type, it'll now end up getting boxed to use the object overload.
+        //   This also necessitates the T? overload, since nullable value types don't meet a T : struct constraint, so without those
+        //   they'd all map to the object overloads as well.
+        // - Any reference type with an implicit cast to ROS<char> will fail to compile due to ambiguities between the overloads. string
+        //   is one such type, hence needing dedicated overloads for it that can be bound to more tightly.
+        //
+        // A middle ground we've settled on, which is likely to be the right approach for most other builders as well, would be the set:
+        //     (T, ...) with no constraint
+        //     (ReadOnlySpan<char>) and (ReadOnlySpan<char>, int)
+        //     (object, int alignment = 0, string? format = null)
+        //     (string) and (string, int)
+        // This would address most of the concerns, at the expense of:
```

Comment:
Do you foresee a world where people start creating large (multi-field) value types and want to pass those types to the builder? Do we want to reserve the right to pass this data byref (via _in_) in the future?

Reply:
> Do you foresee a world where people start creating large (multi-field) value types and want to pass those types to the builder?

Yes, I expect that will happen.  I wouldn't be surprised, for example, if record structs were taught to implement this interface in the future.

> Do we want to reserve the right to pass this data byref (via in) in the future?

@333fred, the compiler allows that, yes?  @GrabYourPitchforks, are you suggesting we change the T overloads now to pass as `in`?

Reply:
> are you suggesting we change the T overloads now to pass as in?

I wouldn't do that right now, since I imagine the common case will be int / long / DateTime / other types which fit cleanly into registers. Passing them byref would be a perf loss. It was more about whether a future compiler would try to be smart and guess whether passing a particular piece of data byref vs byval would be more efficient. I also don't remember offhand how constrained interface dispatch works in the face of _in_ parameters.

Reply:
> whether a future compiler would try to be smart and guess whether passing a particular piece of data byref vs byval would be more efficient

I see.  You're suggesting in the future we could add another overload
```C#
TryFormatInterpolationHole<T>(T value) // today
TryFormatInterpolationHole(in T value) // tomorrow as well
```
and then the compiler could decide which to call based on its knowledge of the size of the struct.

Reply:
> @333fred, the compiler allows that, yes?

I don't believe so. Overload resolution has well-defined rules, I don't believe we can start changing which overload is selected based on the size of the struct.

Reply:
>> the compiler allows that, yes?

> I don't believe so.

My question was about the compiler allowing TryFormatInterpolationHole to have an `in` value.

The rest of the discussion around then overloading based on `in`-ness and enabling the compiler to choose one was future speculation.

Reply:
> My question was about the compiler allowing TryFormatInterpolationHole to have an in value. The compiler / your proposal allows that, yes?

Yes, it should be allowed.